rust/connlib/clients/shared/src/eventloop.rs

```diff
@@ -0,0 +1,467 @@
+                }) => {
+                    let resource_id = resource.id();
+
+                    if let Err(e) = self.tunnel.add_resource(resource) {
+                        tracing::warn!(%resource_id, "Failed to add resource: {e}");
+                    }
+                }
+                Poll::Ready(phoenix_channel::Event::InboundMessage {
+                    msg: IngressMessages::ResourceDeleted(RemoveResource(resource_id)),
+                    ..
+                }) => {
+                    // TODO
+                }
+                Poll::Ready(phoenix_channel::Event::SuccessResponse {
+                    res:
+                        ReplyMessages::Connect(Connect {
+                            gateway_payload:
+                                GatewayResponse::ConnectionAccepted(ConnectionAccepted {
+                                    ice_parameters,
+                                    domain_response,
+                                }),
+                            gateway_public_key,
+                            resource_id,
+                            persistent_keepalive,
+                        }),
+                    ..
+                }) => {
+                    if let Err(e) = self.tunnel.received_offer_response(
+                        resource_id,
+                        ice_parameters,
+                        domain_response,
+                        gateway_public_key.0.into(),
+                    ) {
+                        tracing::warn!("Failed to accept connection: {e}");
+                    }
+                }
+                Poll::Ready(phoenix_channel::Event::SuccessResponse {
+                    res:
+                        ReplyMessages::Connect(Connect {
+                            gateway_payload:
+                                GatewayResponse::ResourceAccepted(ResourceAccepted { domain_response }),
+                            gateway_public_key,
+                            resource_id,
+                            persistent_keepalive,
+                        }),
+                    ..
+                }) => {
+                    if let Err(e) = self
+                        .tunnel
+                        .received_domain_parameters(resource_id, domain_response)
+                    {
+                        tracing::warn!("Failed to accept resource: {e}");
+                    }
+                }
+                Poll::Ready(phoenix_channel::Event::SuccessResponse {
+                    res:
+                        ReplyMessages::ConnectionDetails(ConnectionDetails {
+                            gateway_id,
+                            resource_id,
+                            relays,
+                            gateway_remote_ip,
+                        }),
+                    ..
+                }) => {
+                    match self
+                        .tunnel
+                        .request_connection(resource_id, gateway_id, relays, todo!())
```

Comment:
I never fully understood what we are doing with the `reference` here as part of `request_connection`. It is my understanding the from the perspective of Phoenix LiveView, `reference` is a unique identifier for a request. Thus, in the `phoenix-channel` implementation, this is managed by the library and sending a request gives you an `OutboundRequestId` that you can use to match responses.

For `request_connection`, we seem to be using the reference as some kind of counter? Why do we need this? Can we just remove that?

Reply:
If we talk about `ref` that is pushed by `request_connection` message and then sent back in `connection_ready` then it's our own implementation and we need it to track for which connection the reply is. The protocol has replies for the server but doesn't have them for the clients, so we built them ourselves.

BTW `Phoenix LiveView` is just for front-end, this is all Phoenix PubSub.